compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -313,12 +315,122 @@ getTypeClassDictionary world = \case
     _ ->
         Nothing
 
-simplifyExpr :: World -> Expr -> Expr
-simplifyExpr world = fst . cata go
+-- | Go inside a closed expression and try to inline and beta reduce
+-- types and arguments, to try monomorphizing.
+inlineClosedExpr :: Int -> World -> Expr -> Simplifier Expr
+inlineClosedExpr d world e = do
+    e' <- go e
+    if e == e'
+        then pure e
+        else simplifyExpr' (d-1) e'
+  where
+    go = \case
+        e@(EVal x) -> do
+            case lookupValue x world of
+                Left _ -> pure e
+                Right dval -> pure (dvalBody dval)
```

Comment:
Why are you inlining values? Aren’t you potentially creating duplication here, i.e., the exact opposite of what this transformation is supposed to achieve?

Reply:
The point of this is to propagate constants, like types, dictionaries, and other literals. Otherwise you get very little memoization.

Reply:
I’m not sure I’m following, Let’s say I am lifting `f x` where `f` and `x` are bound at the top-level and they are both fairly large. in that case lifting it at `f x` seems better than inlining `f` and `x`.